query-node/mappings/src/forum.ts

```diff
@@ -82,6 +83,56 @@ async function getPost(store: DatabaseManager, postId: string, relations?: 'thre
   return post
 }
 
+type ModerationType = 'runtime' | 'leadRemark' | 'workerRemark'
+
+async function canModerate(store: DatabaseManager, actorId: string, categoryId: string): Promise<boolean> {
+  const category = await getCategory(store, categoryId, ['moderators', 'parent'])
+  return (
+    category.moderators.some(({ id }) => id === actorId) ||
+    (category.parent ? canModerate(store, actorId, category.parent.id) : false)
+  )
+}
+
+export async function moderatePost(
+  store: DatabaseManager,
+  event: SubstrateEvent,
+  type: ModerationType,
+  postId: string,
+  actor: Worker,
+  rationale: string
+): Promise<void> {
+  const post = await getPost(store, postId, ['thread', 'thread.category'])
```

Comment:
ref my other comment. call to getPost here can throw if unvalidated postId is not found. So wrap this with try/catch and return `invalidMetadata()` on error.